runtimebp/internal/maxprocs/maxprocs.go

```diff
@@ -0,0 +1,129 @@
+}
+
+const (
+	// The default scale applies when request is set but not scale (or scale is
+	// invalid). It defaults to some "oversubscription" to take advantage of
+	// excess Kubernetes node capacity when available.
+	//
+	// Extremely performance-sensitive services should set scale to 1.0 to
+	// ensure consistent performance.
+	//
+	// This value was NOT chosen scientifically and IS subject to change.
+	defaultCPURequestScale = 1.5
+
+	setByGOMAXPROCS   = "gomaxprocs"
+	setByRequest      = "cpu_request"
+	setByAutomaxprocs = "automaxprocs"
+)
+
+var (
+	mEnvGOMAXPROCS      = promauto.NewGaugeVec(prometheus.GaugeOpts{Name: "baseplate_go_env_gomaxprocs"}, []string{"present"})
+	mEnvCPURequest      = promauto.NewGaugeVec(prometheus.GaugeOpts{Name: "baseplate_go_env_baseplate_cpu_request"}, []string{"present"})
+	mEnvCPURequestScale = promauto.NewGaugeVec(prometheus.GaugeOpts{Name: "baseplate_go_env_baseplate_cpu_request_scale"}, []string{"present"})
+
+	initialGOMAXPROCS = promauto.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "baseplate_go_initial_gomaxprocs",
+	}, []string{"set_by"})
+
+	_ = promauto.NewGaugeFunc(prometheus.GaugeOpts{
+		Name: "baseplate_go_current_gomaxprocs",
+	}, currentGOMAXPROCS)
+
+	// overridable for tests
+	setWithAutomaxprocs = func() {
+		ubermaxprocs.Set(
+			ubermaxprocs.Min(2),
+			ubermaxprocs.Logger(func(s string, i ...interface{}) {
+				fmt.Fprintf(os.Stderr, s, i...) // contains context
+			}),
+		)
+	}
+)
+
+// Set configures the runtime's GOMAXPROCS using the following heuristic:
+//   1. If $GOMAXPROCS is set, Set relinquishes control to the Go runtime.
+//      This should cause the runtime to respect this value directly.
+//   2. If $BASEPLATE_CPU_REQUEST is unset/invalid, Set relinquishes control to automaxprocs, minimum 2.
+//      See https://pkg.go.dev/go.uber.org/automaxprocs for specific behavior.
+//   3. Otherwise, $BASEPLATE_CPU_REQUEST is multiplied by $BASEPLATE_CPU_REQUEST_SCALE
+//      (or defaultCPURequestScale) to compute the new GOMAXPROCS, minimum 2.
+//
+// Set also exports several metrics to facilitate further tuning/analysis.
+func Set() {
+	envGOMAXPROCS := &floatEnv{key: "GOMAXPROCS", gauge: mEnvGOMAXPROCS}
+	envCPURequest := &floatEnv{key: "BASEPLATE_CPU_REQUEST", gauge: mEnvCPURequest}
+	envCPURequestScale := &floatEnv{key: "BASEPLATE_CPU_REQUEST_SCALE", gauge: mEnvCPURequestScale}
+
+	for _, env := range []*floatEnv{envGOMAXPROCS, envCPURequest, envCPURequestScale} {
+		env.raw, env.present = os.LookupEnv(env.key)
+		env.val, _ = strconv.ParseFloat(env.raw, 64) // error not needed as we check val > 0 and save raw
+		env.gauge.WithLabelValues(strconv.FormatBool(env.present)).Set(env.val)
+	}
+
+	setBy := setByGOMAXPROCS
+	defer func() {
+		initialGOMAXPROCS.WithLabelValues(setBy).Set(currentGOMAXPROCS())
+	}()
+
+	if envGOMAXPROCS.present {
+		return // let Go runtime handle it
+	}
+
+	if !envCPURequest.present {
+		setBy = setByAutomaxprocs
+		setWithAutomaxprocs()
+		return
```

Comment:
instead of using automaxprocs, just move the unexported functions in `runtimebp` here to read cpu limits from cgroup v1 and v2, and use that info. `runtimebp.NumCPU` can also call those functions and since this is `internal` we don't need to worry about exporting them and adding new APIs.

Reply:
I'd like to keep this implementation as close to baseplate v2 as possible, and we're not going to pull in the runtimebp functions there as automaxprocs nicely encapsulates and maintains this difficult-to-test logic.